### PR TITLE
Servlet to purge old rooms

### PR DESCRIPTION
**Pull request by richvdh**
(I needed this to clear out a big room from my personal HS...)

### CODE REVIEW

File: synapse/rest/admin/purge_room_servlet.py [hunk 1]
```diff
@@ -0,0 +1,58 @@
...
+from synapse.http.servlet import (
+    RestServlet,
+    assert_params_in_dict,
+    parse_json_object_from_request,
+)
+from synapse.rest.admin import assert_requester_is_admin
+
+
+class PurgeRoomServlet(RestServlet):
+    """Servlet which will remove all trace of an a room from the database
```
Comment (anoadragon453) [inline synapse/rest/admin/purge_room_servlet.py]:
```suggestion
    """Servlet which will remove all trace of a room from the database
```

File: synapse/rest/admin/purge_room_servlet.py [hunk 2]
```diff
@@ -0,0 +1,58 @@
...
+
+    POST /_synapse/admin/v1/purge_room
+    {
+        "room_id": "!room:id"
+    }
+
+    returns:
+
+    {
```
Comment (anoadragon453) [inline synapse/rest/admin/purge_room_servlet.py]:
```suggestion
    {}
```

File: synapse/rest/admin/purge_room_servlet.py [hunk 3]
```diff
@@ -0,0 +1,58 @@
...
+    }
```
Comment (anoadragon453) [inline synapse/rest/admin/purge_room_servlet.py]:
```suggestion
```

File: synapse/storage/events.py [hunk 4]
```diff
@@ -2191,6 +2191,89 @@ def _find_unreferenced_groups_during_purge(self, txn, state_groups):
...
+    def _purge_room_txn(self, txn, room_id):
+        # first we have to delete the state groups states
+        logger.info("[purge] removing %s from state_groups_state", room_id)
+
+        txn.execute(
+            "DELETE FROM state_groups_state "
+            "WHERE state_group IN ("
+            "SELECT state_group FROM events JOIN event_to_state_groups USING(event_id) "
+            "WHERE events.room_id=?"
+            ")",
```
Comment (erikjohnston) [inline synapse/storage/events.py]: I really do prefer using multiline strings for this sort of stuff, as it makes it a *lot* easier to then c+p to debug.

**Comment by erikjohnston** [comment]
FTR the full list of tables with columns called `room_id` and `event_id` is generated with:

```sql
select distinct c.relname
from pg_class as c
    inner join pg_attribute as a on a.attrelid = c.oid
where a.attname in ('room_id', 'event_id') and c.relkind = 'r';
```

And gives:

```
 appservice_room_list
 blocked_rooms
 current_state_delta_stream
 current_state_events
 e2e_room_keys
 event_auth
 event_backward_extremities
 event_edges
 event_forward_extremities
 event_json
 event_push_actions
 event_push_actions_staging
 event_push_summary
 event_reference_hashes
 event_relations
 event_reports
 event_search
 event_to_state_groups
 events
 ex_outlier_stream
 group_rooms
 group_summary_rooms
 guest_access
 history_visibility
 local_invites
 public_room_list_stream
 pusher_throttle
 receipts_graph
 receipts_linearized
 redactions
 rejections
 room_account_data
 room_aliases
 room_depth
 room_memberships
 room_names
 room_state
 room_stats
 room_stats_earliest_token
 room_tags
 room_tags_revisions
 rooms
 state_events
 state_groups
 state_groups_state
 stream_ordering_to_exterm
 topics
 user_directory
 users_in_public_rooms
 users_who_share_private_rooms
(50 rows)
```

Looks like you're missing a bunch from room stats work and things.

File: synapse/storage/events.py (hunk 4 above)
Comment (richvdh) [reply synapse/storage/events.py]: fair, yes. For some reason I forget to do this.

**Comment by richvdh** [comment]
Ok, I have added a load of tables to the list. PTAL.

**Pull request merged**